File: landa/landa_sales/sales_order/sales_order.js
// Copyright (c) 2021, Real Experts GmbH and contributors
// For license information, please see license.txt

frappe.ui.form.on("Sales Order",  {
    setup: function (frm) {
        frm.set_query("shipping_contact", erpnext.queries.contact_query);
    },
    refresh: function (frm) {
        if (frm.is_new()) {
            landa.utils.set_company_and_customer(frm);
        }

        frm.set_query("item_code", "items", function(doc) {
            return {
                query: "erpnext.controllers.queries.item_query",
                filters: {
                    valid_from_year: ["<=", doc.year_of_settlement],
                    valid_to_year: [">=", doc.year_of_settlement],
                    cannot_be_ordered: 0,
                    is_sales_item: 1,
                    company: doc.company,
                }
            };
        });

        frm.set_query("selling_price_list", function(doc) {
            return {
                filters: {
                    selling: 1,
                    company: doc.company,
                }
            };
        });

        if (!frappe.user.has_role("LANDA Regional Organization Management")) {
            setTimeout(function() {
                frm.remove_custom_button(__("Hold"), __("Status"));
                frm.remove_custom_button(__("Close"), __("Status"));
            }, 500);
        }

        setTimeout(function() {
            frm.remove_custom_button(__("Pick List"), __("Create"));
            frm.remove_custom_button(__("Work Order"), __("Create"));
            frm.remove_custom_button(__("Material Request"), __("Create"));
            frm.remove_custom_button(__("Request for Raw Materials"), __("Create"));
            frm.remove_custom_button(__("Purchase Order"), __("Create"));
            frm.remove_custom_button(__("Project"), __("Create"));
            frm.remove_custom_button(__("Subscription"), __("Create"));
            frm.remove_custom_button(__("Payment Request"), __("Create"));
            frm.remove_custom_button(__("Payment Request"), __("Create"));
        }, 500);
    },
    before_save: function (frm) {
        frm.doc.items = landa.selling.remove_zero_qty_items(frm.doc.items);
    },
    year_of_settlement: function (frm) {
        landa.selling.prefill_items(frm);
        frm.doc.delivery_date = new Date(frm.doc.year_of_settlement, 11, 31); // the month is 0-indexed
    },
    shipping_contact: function (frm) {
        landa.selling.set_contact_details(frm, "shipping");
    }
});

File: landa/landa_sales/sales_order/sales_order.py
# Copyright (c) 2021, Real Experts GmbH and contributors
# For license information, please see license.txt

import frappe
from erpnext.selling.doctype.sales_order.sales_order import make_delivery_note, make_sales_invoice
from frappe.utils.data import get_year_ending

from landa.landa_sales.utils import (
	validate_company_customer,
	validate_company_price_list,
	validate_year_of_settlement,
)
from landa.utils import update_doc


def before_validate(sales_order, event):
	for item in sales_order.items:
		item.delivery_date = get_year_ending(str(sales_order.year_of_settlement))

	if (not sales_order.tax_category) and frappe.db.exists("Tax Category", "Umsatzsteuer"):
		sales_order.tax_category = "Umsatzsteuer"


def validate(doc, event):
	validate_year_of_settlement(doc)
	validate_company_customer(doc.company, doc.customer)
	validate_company_price_list(doc.company, doc.selling_price_list)


def autoname(doc, event):
	"""Create Company-specific Sales Order name."""
	from landa.utils import get_new_name

	doc.name = get_new_name("BEST", doc.company, "Sales Order", doc.year_of_settlement)


def get_dashboard_data(data):
	data["transactions"] = [
		{
			"label": "Sales",
			"items": ["Delivery Note", "Sales Invoice"],
		},
	]
	return data


@frappe.whitelist()
def make_landa_sales_invoice(source_name, target_doc=None, ignore_permissions=False):
	source_doc = frappe.get_doc("Sales Order", source_name)
	target_doc = make_sales_invoice(source_name, target_doc, ignore_permissions)

	update_doc(source_doc, target_doc)

	return target_doc


@frappe.whitelist()
def make_landa_delivery_note(source_name, target_doc=None, skip_item_mapping=False):
	source_doc = frappe.get_doc("Sales Order", source_name)
	target_doc = make_delivery_note(source_name, target_doc, skip_item_mapping)

	update_doc(source_doc, target_doc)

	return target_doc
